### ISSUE TITLE
Alignment step fails: short of disk space or pipe issues?

### ISSUE THREAD

**Issue by naumenko-sa** [bcbio/bcbio-nextgen#3120]
Hello everyone!

Mostly to document this issue, but I'd also appreciate any thoughts.

I'm fine tunning fgbio parameters for https://github.com/bcbio/bcbio-nextgen/blob/master/config/templates/somatic-single-umi.yaml

Input files are high coverage gene panels, 27G+ 27G paired end fastq.

The command:
```
bwa mem  \
-c 250 \
-M \
-t 16 \
-R '@RG\tID:sample_name\tPL:illumina\tPU:sample_name\tSM:sample_name' \
-v 1 \
/path/sample_name_R1.fq.gz \
/path/sample_name_R2.fq.gz  | \
bamsormadup \
tmpfile=./sort-sorttmp-markdup inputformat=sam \
threads=16 \
outputformat=bam \
level=0 \
SO=coordinate | \
python /path/anaconda/bin/umis bamtag - | \
samtools view -b > sample_name-sort.bam
```
fails with:
```
Traceback (most recent call last):
  File "/data/bcbio/galaxy/../anaconda/bin/umis", line 11, in <module>
    load_entry_point('umis==1.0.6', 'console_scripts', 'umis')()
  File "/data/bcbio/anaconda/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/data/bcbio/anaconda/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/data/bcbio/anaconda/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/data/bcbio/anaconda/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/data/bcbio/anaconda/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/data/bcbio/anaconda/lib/python3.6/site-packages/umis/umis.py", line 1219, in bamtag
    sam_file = open_bamfile(sam)
  File "/data/bcbio/anaconda/lib/python3.6/site-packages/umis/umis.py", line 123, in open_bamfile
    return AlignmentFile(sam, mode=sam_mode)
  File "pysam/libcalignmentfile.pyx", line 741, in pysam.libcalignmentfile.AlignmentFile.__cinit__
  File "pysam/libcalignmentfile.pyx", line 990, in pysam.libcalignmentfile.AlignmentFile._open
ValueError: file has no sequences defined (mode='rb') - is it SAM/BAM format? Consider opening with check_sq=False
[M::mem_pestat] skip orientation FF as there are not enough pairs
[M::mem_pestat] analyzing insert size distribution for orientation FR...
[M::mem_pestat] (25, 50, 75) percentile: (161, 176, 301)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 581)
[M::mem_pestat] mean and std.dev: (222.93, 88.14)
[M::mem_pestat] low and high boundaries for proper pairs: (1, 721)
[M::mem_pestat] skip orientation RF as there are not enough pairs
[M::mem_pestat] skip orientation RR as there are not enough pairs
' returned non-zero exit status 1.
```

I decomposed the pipe, and ran 4 steps:
```
1. bwa mem, result = test.sam, 360G
2. bamsormadup, result = test.bam, 306G
3. umis bamtag, result = test1.sam, 396G
4. samtools view -b, result = test1.bam, 34G
```
They finished ok, so I think the issue is either with free disk space (the intermediate files are huge, and it is 2T HDD AWS instance), or something is wrong with piping of big files?
```
ulimit -a
core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 251707
max locked memory       (kbytes, -l) 16384
max memory size         (kbytes, -m) unlimited
open files                      (-n) 100000
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 251707
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited
```

I'm trying to give bcbio more space. 

As a side note, we are using biobambam2 from the latest binary release on github:
https://github.com/gt1/biobambam2/releases, v.2.0.87.

There was some progress in development and bug fixing in biobambam2 and a new binary release available:
https://gitlab.com/german.tischler/biobambam2/-/tags/2.0.106-release-20190614124427 
We could update its bioconda recipe to pull the new one:
https://github.com/bioconda/bioconda-recipes/blob/master/recipes/biobambam/meta.yaml

SN

**Comment by naumenko-sa** [comment]
upd:

not a free disk space issue.
https://github.com/vals/umis/blob/master/umis/umis.py#L123 ?
but why it is working step by step?

**Comment by naumenko-sa** [comment]
upd:

- random failures in [34min, 1h]
- biobambam2 2.0.16 - same issue

umis issue?

**Comment by matthdsm** [comment]
Hi Sergey,

while you're decomposing the workflow, you might want to check out [Elprep](https://github.com/ExaScience/elprep). Does exactly the same thing, but should be faster. Keep in mind though it's quite MEM hungry.
Perhaps this might fix the issues?
Cheers
M

**Comment by naumenko-sa** [comment]
thanks @matthdsm!
I have not tried Elprep yet. 

I tried samtools instead of biobambam2:

```
bwa mem \
-c 250 -M -t 16  -R '@RG\tID:sample_name\tPL:illumina\tPU:sample_name\tSM:sample_name' \
-v 1 /data/bcbio/genomes/Hsapiens/hg38/bwa/hg38.fa \
/path/sample_name_R1.fq.gz \
/path/sample_name_R2.fq.gz | \
samtools sort -@ 16 -n -O bam /dev/stdin |
samtools fixmate -@ 16 -O bam /dev/stdin /dev/stdout |
samtools sort -@ 16 -O bam -T sample-sort_tmp.bam /dev/stdin | \
samtools markdup -@ 16 -O bam -T sample-dup_tmp.bam /dev/stdin /dev/stdout | \
/data/bcbio/anaconda/bin/python /data/bcbio/galaxy/../anaconda/bin/umis bamtag - | samtools view -b > output.bam
```

failes with
```
[main] Real time: 13299.637 sec; CPU: 196642.504 sec
[bam_sort_core] merging from 400 files and 16 in-memory blocks...
[bam_sort_core] merging from 400 files and 16 in-memory blocks...
[markdup] error: no ms score tag. Please run samtools fixmate on file first.
[markdup] error: no ms score tag. Please run samtools fixmate on file first.
Traceback (most recent call last):
  File "/data/bcbio/galaxy/../anaconda/bin/umis", line 11, in <module>
    load_entry_point('umis==1.0.6', 'console_scripts', 'umis')()
  File "/data/bcbio/anaconda/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/data/bcbio/anaconda/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/data/bcbio/anaconda/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/data/bcbio/anaconda/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/data/bcbio/anaconda/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/data/bcbio/anaconda/lib/python3.6/site-packages/umis-1.0.6-py3.6-linux-x86_64.egg/umis/umis.py", line 1225, in bamtag
    queryalignment = next(track)
  File "pysam/libcalignmentfile.pyx", line 2197, in pysam.libcalignmentfile.IteratorRowAll.__next__
StopIteration
```
All these experiment were on AWS instance 
```
Ubuntu 4.15.0-1060-aws
```

on the cluster the same run just runs with no issues (CentOS7)

@roryk @chapmanb  any ideas?

**Comment by chapmanb** [comment]
Sergey;
Sorry about all the issues. My guess is that one of the upstream processes in the pipe like alignment hits memory issues, spits out an error further up buried in the traceback, and this results in a badly formatted file downstream triggering the error you see.

If the same process runs fine locally, my guess would be memory problems on AWS. Some of the instances don't have a lot of memory per core. 4Gb+/core is normally the safe spot for this. bwa doesn't actually control memory size in a consistent way so lower memory with the same number of cores can trigger issues due to both the alignment and all the downstream processing happening at the same time.

If it possible to try on a machine with more memory/core to see if that helps resolve the problem? Hope this helps,
Brad

**Comment by naumenko-sa** [comment]
Thanks Brad for the suggestions, this helped me to connect this story with the other one where we profiled memory/CPU and also saw some random failures. 

Here I tried 4 Cores with 64G RAM, i.e. 16G/core for bwa mem. Failed as well after 1h14. Trying with 2 cores.

Another user reported the same issue, but she reran on another AWS instance and it was fine.

This reminds me one sample I had trouble processing before which had a spike of coverage in one particular locus causing memory issues.

So for now the hypothesis is memory issue, i.e. when running a pipe, all commands eat RAM, and bwa gets less RAM and fails, and when bwa is ran alone, it fits in the RAM. 

Practically, I'm going to profile RAM to confirm the issue, and then profile coverage to see the problematic locus.

Thanks again for pointing in the right direction!
SN
